pkg/vsphere/vm/vm.go

```diff
@@ -712,3 +713,20 @@ func (vm *VirtualMachine) RemoveSnapshotByRef(ctx context.Context, snapshot *typ
 
 	return object.NewTask(vm.Vim25(), res.Returnval), nil
 }
+
+// IsAlreadyPoweredOffError is an accessor method because of the number of times package name and
+// variable name tend to collide for VMs.
+func (vm *VirtualMachine) IsAlreadyPoweredOffError(err error) bool {
+	return IsAlreadyPoweredOffError(err)
+}
+
+// IsAlreadyPoweredOffError verifies that the error is an InvalidPowerState error and
+// returns true if the existing state from the error is powered off
+func IsAlreadyPoweredOffError(err error) bool {
+	if soap.IsSoapFault(err) {
+		invalidState, ok := soap.ToSoapFault(err).VimFault().(types.InvalidPowerState)
```

Comment:
are you sure it is a type.InvalidePowerState, not a pointer to it?
There is a piece of code in a portlayer that operates with a pointer:
```Go
func (c *containerBase) poweroff(op trace.Operation) error {
	// make sure we have vm
	if c.vm == nil {
		return NotYetExistError{c.ExecConfig.ID}
	}

	_, err := c.vm.WaitForResult(op, func(op context.Context) (tasks.Task, error) {
		return c.vm.PowerOff(op)
	})

	if err != nil {

		// It is possible the VM has finally shutdown in between, ignore the error in that case
		if terr, ok := err.(task.Error); ok {
			switch terr := terr.Fault().(type) {
			case *types.InvalidPowerState:
				if terr.ExistingState == types.VirtualMachinePowerStatePoweredOff {
					op.Warnf("power off %s task skipped (state was already %s)", c, terr.ExistingState)
					return nil
				}
				op.Warnf("invalid power state during power off: %s", terr.ExistingState)

			case *types.GenericVmConfigFault:

				// Check if the poweroff task was canceled due to a concurrent guest shutdown
				if len(terr.FaultMessage) > 0 {
					k := terr.FaultMessage[0].Key
					if k == vmNotSuspendedKey || k == vmPoweringOffKey {
						op.Infof("power off %s task skipped due to guest shutdown", c)
						return nil
					}
				}
				op.Warnf("generic vm config fault during power off: %#v", terr)

			default:
				op.Warnf("hard power off failed due to: %#v", terr)
			}
		}

		return err
	}

	return nil
}
```